File: Prime-numbers/sieve.py

'''
Consider that you are reading a number n from the file sieve.in
This script factorises all of the numbers less or equal to n into prime factors.
Enjoy! :D 

Copyright Tudor Coman 2015. All rights reserved.

'''
sieve = [1, 1]

fout = open("sieve.out", "wt")

def next_int(fname):
    for line in open(fname):
        for val in line.split():
            yield int(val)

def simple_sieve(nr):
	i = 4
	while (i < nr + 1):
		sieve[i] = 1
		i += 2
	i = 3
	while(i * i < nr + 1):
		if sieve[i] == 0:
			j = i * i
			while(j < nr + 1):
				sieve[j] = 1
				j += 2 * i
		i += 2



def special_sieve(nr):
	i = 4
	while (i < nr + 1):
		sieve[i] = 2
		i += 2
	i = 3
	while(i * i < nr + 1):
		if sieve[i] == 0:
			j = i * i
			while(j < nr + 1):
				sieve[j] = i
				j += 2 * i
		i += 2

def prime_factorisation(nr):

	if nr == 1:
		fout.write("\n")
		return

	if sieve[nr] == 0:
		fout.write("%d^1\n" % (nr)) 
		return

	div = sieve[nr]
	exp = 0
	no = nr
	while (nr % div == 0):
		nr /= div
		exp += 1

	fout.write("%d^%d " % (div, exp)) 
	prime_factorisation(nr)

for n in next_int("sieve.in"):
	''' Initialize the sieve vector'''
	for i in range(2, n + 1):
		sieve.append(0)

	special_sieve(n)
	for i in range(2, n + 1):
		fout.write("%d: " % (i)) 
		prime_factorisation(i)

fout.close()

File: Prime-numbers/sieve.cpp

#include <fstream>
#include <iostream>
using namespace std;

const int MAX_N = 500000;
int sieve[MAX_N + 1];

/*** CLASSIC SIEVE ***/
void simple_sieve(int nr) {
	
	sieve[0] = sieve[1] = 1;

	for(register int i = 4; i <= nr; i += 2)
		sieve[i] = 1;

	for(register int i = 3; 1LL * i * i <= nr; i += 2)
		if(!sieve[i])
			for(register int j = i * i; j <= nr; j += (i << 1))
				sieve[j] = 1;

}

/*** SPECIAL SIEVE THAT GENEREATES THE BIGGEST PRIME DIVISOR OF EACH NUMBER ***/
void special_sieve(int nr) {

	sieve[0] = sieve[1] = 1;

	for(register int i = 4; i <= nr; i += 2)
		sieve[i] = 2;

	for(register int i = 3; 1LL * i * i <= nr; i += 2)
		if(sieve[i] == 0)
			for(register int j = i * i; j <= nr; j += (i << 1))
				sieve[j] = i;

}

void prime_factorisation(int nr) {
	if(nr == 1) {
		cout << '\n';
		return;
	}

	if(sieve[nr] == 0) {
		cout << nr << "^" << 1 << '\n';
		return;
	}
		
	int div = sieve[nr], exp = 0, no = nr;

	while(nr % div == 0)
		nr /= div, ++ exp;

	cout << div << '^' << exp << ' ';

	prime_factorisation(nr);

}

int main(void) {
	
	ifstream f("sieve.in");
	int N;
	f >> N;
	special_sieve(N);

	for(register int i = 2; i <= N; ++ i)
		cout << i << ':', prime_factorisation(i);
	return 0;

}
